### PR TITLE
Feature/custom validator

### PR DESCRIPTION
**Pull request by bentpr**
Number of changes to support custom validation attributes

### CODE REVIEW

File: GovUk.Frontend.AspNetCore.Extensions/README.md [hunk 1]
```diff
@@ -197,3 +197,91 @@ If enabled, the extension library automatically generates localised strings for
...
+
+        validator.addMethod('banana', function (value, element, params) {
+            var additionalParameter = params.additionalParameter;
+            // Do some validation here            
+            return true;
+        });
+
+        validator.unobtrusive.adapters.add("banana", ["additionalParameter"],
+            function (options) {
+                options.rules['banana'] = {
```
Comment (sussexrick) [inline GovUk.Frontend.AspNetCore.Extensions/README.md]: could be `options.rules['banana'] = options.params`

File: GovUk.Frontend.AspNetCore.Extensions/README.md [hunk 2]
```diff
@@ -197,3 +197,91 @@ If enabled, the extension library automatically generates localised strings for
...
+                };
+                options.messages['banana'] = options.message;
+            }
+        );
+        validator.unobtrusive.parse();
+    }
+</script>
+```
+
+3. All being well, you should find that the generated html element now has `data-val-banana` and `data-val-banana-additionaparamter` attributes. These correspond to the validation attribute added to the view model.
```
Comment (sussexrick) [inline GovUk.Frontend.AspNetCore.Extensions/README.md]: typos in `additionalparameter`

File: GovUk.Frontend.AspNetCore.Extensions/Validation/ClientSideValidationHtmlEnhancer.cs [hunk 3]
```diff
@@ -243,6 +248,38 @@ private static bool ErrorMessageTagHelperHasRenderedAnError(HtmlNode? errorMessa
...
+                                foreach (var attr in attrDictionary)
+                                {
+                                    targetElement.Attributes.Add(attr.Key, attr.Value);
+                                }
+                            }
+                            validateElement = true;
+                        }
+                    }
+
                     if (validateElement) { targetElement.Attributes.Add("data-val", "true"); }
```
Comment (sussexrick) [inline GovUk.Frontend.AspNetCore.Extensions/Validation/ClientSideValidationHtmlEnhancer.cs:297]: If `customValidation.AddValidation()` has added `data-val=true` would we end up with a duplicate attribute here?

File: GovUk.Frontend.AspNetCore.Extensions/__tests__/govuk-validation-validateCustom.tests.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        _mockValidator.methods["custom"].mockReturnValue(false);
+
+        const input = document.querySelector("input");
+        testSubject.validateElement(input);
+
+        expect(_mockUpdateError.mock.calls.length).toBe(1);
+        expect(_mockUpdateError.mock.calls[0][0]).toBe(input);
+        expect(_mockUpdateError.mock.calls[0][1]).toBe(
+            input.getAttribute("data-val-custom")
+        );
```
Comment (sussexrick) [inline GovUk.Frontend.AspNetCore.Extensions/__tests__/govuk-validation-validateCustom.tests.js:33]: Not testing that the custom property is passed and set to Banana

File: GovUk.Frontend.AspNetCore.Extensions/wwwroot/govuk/govuk-validation.js [hunk 5]
```diff
@@ -261,8 +261,61 @@ function createGovUkValidator() {
     },
 
     validateElement: function (element) {
-      const validator = govuk.getValidator(element.form);
+        const validator = govuk.getValidator(element.form);
 
+        var data = {};
```
Comment (sussexrick) [inline GovUk.Frontend.AspNetCore.Extensions/wwwroot/govuk/govuk-validation.js]: No reason to use `var` in JS, always `let` or `const`

File: GovUk.Frontend.AspNetCore.Extensions/wwwroot/govuk/govuk-validation.js [hunk 6]
```diff
@@ -261,8 +261,61 @@ function createGovUkValidator() {
...
+                    );
+                    if (!valid) {
+                        govuk.updateError(element, value);
+                        break;  // Break so that the errors are displayed in the order they're specified
+                    }
+                }
+            }
+        }
+        
+        /*
```
Comment (sussexrick) [inline GovUk.Frontend.AspNetCore.Extensions/wwwroot/govuk/govuk-validation.js]: This should come out as it's been replaced. We have git history to refer to it.

File: GovUk.Frontend.ExampleApp/Models/CustomValidationViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        [Range(1, 10, ErrorMessage = "Must be between 1 and 10")]
+        public int Field2 { get; set; }
+
+
+        [CustomValidator("Field1", "Field2", ErrorMessage = "This field must be the sum of box 1 and box 2")]
+        public string Field3 { get; set; }
+    }
+
+    [AttributeUsage(AttributeTargets.Property)]
+    public class CustomValidatorAttribute : ValidationAttribute, IClientModelValidator
```
Comment (sussexrick) [inline GovUk.Frontend.ExampleApp/Models/CustomValidationViewModel.cs]: I know you had these together for convenient testing, but separate them out for the PR

File: GovUk.Frontend.ExampleApp/Views/CustomValidation/Index.cshtml [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        govuk.createErrorSummary();
+
+        if (validator) {
+            validator.setDefaults({
+                highlight: govuk.showError,
+                unhighlight: govuk.removeOrUpdateError,
+            });
+
+            validator.addMethod('custom', function (value, element, params) {
+                console.log(value);
```
Comment (sussexrick) [inline GovUk.Frontend.ExampleApp/Views/CustomValidation/Index.cshtml]: These can come out

File: GovUk.Frontend.ExampleApp/Views/CustomValidation/Index.cshtml [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                        return false;
+                    }
+                }
+
+                return true;
+            });
+
+            validator.unobtrusive.adapters.add("custom", ["property1", "property2"],
+                function (options) {
+                    options.rules['custom'] = {
```
Comment (sussexrick) [inline GovUk.Frontend.ExampleApp/Views/CustomValidation/Index.cshtml]: Again I think you can condense this to `options.rules['custom'] = options.params`

File: GovUk.Frontend.ExampleApp/Views/CustomValidation/Index.cshtml [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+                </govuk-input>
+            </govuk-client-side-validation>
+        </govuk-fieldset>
+
+        <govuk-button type="submit">Submit</govuk-button>
+
+    </form>
+    @section scripts {
+    <partial name="GOVUK/Validation" />
+    <script type="text/javascript">
```
Comment (sussexrick) [inline GovUk.Frontend.ExampleApp/Views/CustomValidation/Index.cshtml]: Include the script as a separate file to show good practice. Implemented apps should ban inline JS using a CSP.

**Comment by sussexrick** [comment]
I'd like to see a better error than this in the example app:

image

It's not related to this PR but this page should have the form legend as the h1, like checkboxes do.

image

This one is not working:

image

**Comment by sussexrick** [comment]
On the text input page on the Umbraco app. I suspect this is the length validator showing the wrong message, because there is a minimum length of 5 characters and a value of `12345` does not display this error but `abcde` does.

image

**Review by sussexrick — CHANGES_REQUESTED** [review]

File: GovUk.Frontend.AspNetCore.Extensions/Validation/ClientSideValidationHtmlEnhancer.cs (hunk 3 above)
Comment (bentpr) [reply GovUk.Frontend.AspNetCore.Extensions/Validation/ClientSideValidationHtmlEnhancer.cs:297]: Don't believe so - I've not seen any duplicates?

File: GovUk.Frontend.AspNetCore.Extensions/DataAnnotationsStringLocalizer.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+                return localizedString;
+            }
+        }
+
+        public IEnumerable<LocalizedString> GetAllStrings(bool includeParentCultures)
+            => primaryLocalizer.GetAllStrings(includeParentCultures).Concat(fallbackLocalizer.GetAllStrings(includeParentCultures));
+
+
+        //public IStringLocalizer WithCulture(CultureInfo culture)
```
Comment (sussexrick) [inline GovUk.Frontend.AspNetCore.Extensions/DataAnnotationsStringLocalizer.cs]: An old experiment?

**Review by sussexrick — DISMISSED** [review]

**Review by sussexrick — APPROVED** [review]

**Pull request merged**